test/jdk/java/net/httpclient/HttpResponseConnectionLabelTest.java

```diff
@@ -0,0 +1,446 @@
+            } catch (Exception serverException) {
+                throw new RuntimeException("failed closing server", serverException);
+            }
+        }
+
+        private static ServerRequestPair of(Version version, boolean secure) {
+
+            // Create the server and the request URI
+            SSLContext sslContext = secure ? SSL_CONTEXT : null;
+            HttpTestServer server = createServer(version, sslContext);
+            String handlerPath = "/" + /* salting the path: */ HttpResponseConnectionLabelTest.class.getSimpleName();
+            String requestUriScheme = secure ? "https" : "http";
+            URI requestUri = URI.create(requestUriScheme + "://" + server.serverAuthority() + handlerPath);
+
+            // Register the request handler
+            String serverId = "" + SERVER_COUNTER.getAndIncrement();
+            CountDownLatch[] serverResponseLatchRef = new CountDownLatch[1];
+            server.addHandler(
+                    (exchange) -> {
+                        String responseBody = "" + SERVER_RESPONSE_COUNTER.getAndIncrement();
+                        String connectionKey = exchange.getConnectionKey();
+                        try (exchange) {
+
+                            // Participate in the latch count down
+                            if (serverResponseLatchRef[0] != null) {
+                                serverResponseLatchRef[0].countDown();
+                                LOGGER.log(
+                                        "Server[%s] is waiting for the latch... (connectionKey=%s, responseBody=%s)",
+                                        serverId, connectionKey, responseBody);
+                                serverResponseLatchRef[0].await();
+                            }
+
+                            // Write the response
+                            LOGGER.log(
+                                    "Server[%s] is responding... (connectionKey=%s, responseBody=%s)",
+                                    serverId, connectionKey, responseBody);
+                            exchange.getResponseHeaders().addHeader(CONNECTION_KEY_HEADER_NAME, connectionKey);
+                            exchange.getResponseHeaders().addHeader(SERVER_ID_HEADER_NAME, serverId);
+                            byte[] responseBodyBytes = responseBody.getBytes(CHARSET);
+                            exchange.sendResponseHeaders(200, responseBodyBytes.length);
+                            exchange.getResponseBody().write(responseBodyBytes);
+
+                        } catch (IOException ioe) {
+                            synchronized (ServerRequestPair.class) {
+                                LOGGER.log(
+                                        "Server[%s] has failed! (connectionKey=%s, responseBody=%s)",
+                                        serverId, connectionKey, responseBody);
+                                ioe.printStackTrace(System.out);
+                            }
```

Comment:
Shouldn't we be rethrowing these exceptions from the handler to fail the request?

Reply:
In 4db85c315f698c18a77545d5a2e732edf3199292, removed the `synchronized { ... }` and re-threw the exception.